### ISSUE TITLE
Feature request: Add hardware acceleration for FFmpeg

### ISSUE THREAD

**Issue by Klench** [C4illin/ConvertX#190]
I have just started using this software and it seems great. However, using FFmpeg for conversion of large video files can take a lot of time if using CPU only for transcoding. Using e.g. Intel Quick Sync for hardware acceleration would probably speed things up a lot.

**Comment by C4illin** [comment]
What is missing to make it possible?

**Comment by C10udburst** [comment]
Isn't passing the `/dev/dri` to the container enough?

**Comment by Klench** [comment]
I am not an expert but I assume some hardware acceleration related arguments need to be added to the ffmpeg command:

https://trac.ffmpeg.org/wiki/Hardware/QuickSync

**Comment by C4illin** [comment]
I added FFMPEG_ARGS as en env variable, set it to "-hwaccel qsv" and pass the gpu to the container. Hopefully it works, I dont have an intel cpu but I tested with vulkan.

**Comment by C4illin** [comment]
> Hardware acceleration is working fine in other containers with the same setup.

Does ffmpeg work with hardware acceleration outside the container? Maybe some more args is needed

**Comment by broizter** [comment]
> > Hardware acceleration is working fine in other containers with the same setup.
> 
> Does ffmpeg work with hardware acceleration outside the container? Maybe some more args is needed

It should be enough I think. This works outside the container `ffmpeg -hwaccel qsv -c:v h264_qsv -i BigBuckBunny.mp4 -c:v h264_qsv -b:v 5M test.mp4`

**Comment by C4illin** [comment]
Ok, if the command works by itself but not in the container with the same input there is only two possibilities: you added the wrong device or the alpine version of ffmpeg doesn't inclue qsv

**Comment by C4illin** [comment]
Maybe try again and add `-loglevel verbose -hwaccel_device /dev/dri/renderD128`

**Comment by C4illin** [comment]
You could also test `-hwaccel vaapi`

**Comment by ektorasdj** [comment]
Hello Everyone.

To make it work on my synology with intel celeron J4125 that uses i965 intel video driver (vaapi):
I created a custom dockerfile to install the missing dependencies for the i965 driver to work with Convertx image : 

`FROM ghcr.io/c4illin/convertx:latest

# Install additional dependencies for VAAPI support
RUN apk add --no-cache \
    gcompat \
    libva-utils 

EXPOSE 3000
ENTRYPOINT [ "bun", "run", "./src/index.tsx" ]
`
I also downloaded the i965 driver file : i965_drv_video.so and mapped it inside the container.

Compose file:

`services:
  convertx: 
    image: custom-convertx:latest #the custom image i build with the dockerfile above
    container_name: convertx
    restart: always
    ports:
      - "3050:3000"
    devices:
      - /dev/dri:/dev/dri
    environment:
      - JWT_SECRET=Secret
      - ACCOUNT_REGISTRATION=false
      - HTTP_ALLOWED=true
      - AUTO_DELETE_EVERY_N_HOURS=24
      - FFMPEG_ARGS=-hwaccel vaapi
      - LIBVA_DRIVER_NAME=i965
      - LIBVA_DRIVERS_PATH=/usr/lib/x86_64-linux-gnu/dri/
    volumes:
      - /volume2/docker/convertx/data:/app/data
      - /volume2/docker/convertx/driver:/usr/lib/x86_64-linux-gnu/dri #here is the i965_drv_video.so driver file
    privileged: true #needed for container to have access to the gpu device of the host`

This worked with my synology and converts with hardware accelerated ffmpeg just fine.
If i had to make a request is to include in future images the required libraries and files for vaapi to work with a custom docker image.

In any case i would like to say thank you to the developer since its a great and handy tool!

**Comment by wesleykirkland** [comment]
Adding future notes for people, I used this to get ConvertX working on Intel UHD 650 along with finding the driver file from [here](https://dl-cdn.alpinelinux.org/alpine/v3.20/community/x86_64/intel-media-driver-23.4.3-r0.apk).

```
FROM ghcr.io/c4illin/convertx:latest

# Install additional dependencies for VAAPI support and intel Media driver for UHD Graphics 630
RUN apk add --no-cache \
    intel-media-driver

EXPOSE 3000
ENTRYPOINT [ "bun", "run", "./src/index.tsx" ]
```

```
environment: # Defaults are listed below. All are optional.
      - FFMPEG_ARGS=-hwaccel vaapi # hardware acceleration
      - LIBVA_DRIVER_NAME=iHD # https://github.com/C4illin/ConvertX/issues/190
      - LIBVA_DRIVERS_PATH=/usr/lib/x86_64-linux-gnu/dri
```

**Comment by thecodygriffin** [comment]
Would it be possible to have `libva-utils` and/or the `intel-media-driver` added to the repository image so that a custom Dockerfile doesn't have to be built?

I have to suspect that using integrated Intel GPUs for vaapi hardware acceleration is pretty common.

**Comment by C4illin** [comment]
Yes just send a PR

**Comment by thecodygriffin** [comment]
After some further testing, I found that as long as the host system has the proper drivers installed, the current 0.15.1 repository image does not need any further modification.

For context: I run ConvertX on a Debian "Trixie" 13.1 host with the intel-media-va-driver-non-free installed using Docker Compose. I am using an Intel Core i5-10400 CPU with integrated Intel UHD Graphics 630.

My compose.yml file is as follows:
```
services:
  convertx:
    image: ghcr.io/c4illin/convertx:v0.15.1
    user: 157:1012 # ADDED (optional): Ensure the docker process runs rootless - UID:GID
    group_add: # ADDED (required): Ensure the docker process has permission to access the graphics device
      - 155 # The render group GID on the host
    container_name: convertx
    restart: unless-stopped
    environment:
      - TZ=America/New_York
      - JWT_SECRET=${JWT_SECRET} # will use randomUUID() if unset
      - HTTP_ALLOWED=true # uncomment this if accessing it over a non-https connection
      - ACCOUNT_REGISTRATION=false
      - ALLOW_UNAUTHENTICATED=false
      - AUTO_DELETE_EVERY_N_HOURS=168
      - FFMPEG_ARGS=-hwaccel vaapi # ADDED (required): Enables VAAPI hardware acceleration
      - LIBVA_DRIVER_NAME=iHD # ADDED (required): Informs libva of the intel media driver name
      - HIDE_HISTORY=false
      - LANGUAGE=en
      - UNAUTHENTICATED_USER_SHARING=false
      - MAX_CONVERT_PROCESS=0
    volumes:
      - /mnt/Technology/Containers/convertx/data:/app/data
      - /usr/lib/x86_64-linux-gnu/dri:/usr/lib/x86_64-linux-gnu/dri:ro # ADDED (required): Provides read-only access to the intel media drivers
      - /usr/lib/x86_64-linux-gnu/libigdgmm.so.12:/usr/lib/x86_64-linux-gnu/libigdgmm.so.12:ro # ADDED (required): Provides read-only access to the intel graphics memory management library
    devices: # ADDED (required): Informs container of the graphics device
      - /dev/dri:/dev/dri # The graphics device location mapped from host to container
    networks:
      - convertx_net

networks:
  convertx_net:
    external: true
```
In a nutshell: I believe that as long as you provider the container proper access to the host's graphics device and drivers, you can easily enable hardware acceleration for an integrated Intel GPU as of 0.15.1.

**Comment by aryasenna** [comment]
Hi, is it only me or GPU acceleration in ConvertX with VAAPI/QSV is limited to **decode**?

I see here `-c:v libx264` is **hardcoded**. It means always use SW encoding.

https://github.com/C4illin/ConvertX/blob/f747a5b94be987a57ea3c5f1ce15be16bc0a6d62/src/converters/ffmpeg.ts#L716-L729

That hardcoded value becomes part of `extraArg` and overrides whatever **encoder** setting you put in `FFMPEG_OUTPUT_ARGS=`, since `extraArg` is appended last.

~~I can make a PR that proposes a quick fix, but I'm trying to confirm first if my understanding is correct.~~
Never mind, there's already an open PR addressing this: https://github.com/C4illin/ConvertX/pull/525
